File: sea-shell-service-spring-web-flux/sea-shell-rest-service/src/main/java/org/jesperancinha/shell/webflux/service/SeaShellReactiveServiceImpl.java
package org.jesperancinha.shell.webflux.service;

import org.jesperancinha.shell.client.shells.Shell;
import org.jesperancinha.shell.webflux.data.SeaShellCostumeDto;
import org.jesperancinha.shell.webflux.data.SeaShellDto;
import org.jesperancinha.shell.webflux.data.SeaShellLowerDto;
import org.jesperancinha.shell.webflux.data.SeaShellPersonDto;
import org.jesperancinha.shell.webflux.data.SeaShellTopDto;
import org.jesperancinha.shell.webflux.repo.ShellAccountRepository;
import org.jesperancinha.shell.webflux.repo.ShellCostumeRepository;
import org.jesperancinha.shell.webflux.repo.ShellLowerRepository;
import org.jesperancinha.shell.webflux.repo.ShellPersonRepository;
import org.jesperancinha.shell.webflux.repo.ShellRepository;
import org.jesperancinha.shell.webflux.repo.ShellTopRepository;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.publisher.ParallelFlux;
import reactor.core.scheduler.Schedulers;

import static reactor.core.publisher.Mono.from;
import static reactor.core.publisher.Mono.fromCallable;
import static reactor.core.publisher.Mono.just;
import static reactor.core.publisher.Mono.zip;
import static reactor.core.scheduler.Schedulers.parallel;

@Service
public class SeaShellReactiveServiceImpl extends SeaShellOneAdapter implements SeaShellReactiveService {


    public SeaShellReactiveServiceImpl(ShellRepository shellRepository,
                                       ShellPersonRepository shellPersonRepository,
                                       ShellCostumeRepository shellCostumeRepository,
                                       ShellAccountRepository shellAccountRepository,
                                       ShellTopRepository shellTopRepository,
                                       ShellLowerRepository shellLowerRepository) {
        super(
                shellRepository,
                shellPersonRepository,
                shellCostumeRepository,
                shellAccountRepository,
                shellTopRepository,
                shellLowerRepository
        );
    }

    public ParallelFlux<SeaShellDto> getAllSeaShells() {
        return shellRepository
                .findAllSeaShells()
                .map(this::fetchSeaShellPublisher).flatMap(Flux::from);
    }

    public Mono<SeaShellDto> getShell(final Long id) {
        return shellRepository.findSeaShellById(id).map(this::fetchSeaShellPublisher).flatMap(Mono::from);
    }

    public Mono<Pair<SeaShellPersonDto, SeaShellCostumeDto>> getRootShell(Long idPerson, Long idCostume) {
        return zip(
                shellPersonRepository.findPersonById(idPerson).map(SeaShellConverter::toShellPersonDto).subscribeOn(parallel()),
                shellCostumeRepository.findCostumeById(idCostume).map(SeaShellConverter::toShellCostumeDto).subscribeOn(parallel()),
                Pair::of).subscribeOn(parallel());
    }

    public Mono<Pair<SeaShellTopDto, SeaShellLowerDto>> getRootCostume(Long idTop, Long idLower) {
        return zip(
                shellTopRepository.findTopById(idTop).map(SeaShellConverter::toTopDto).subscribeOn(parallel()),
                shellLowerRepository.findLowerById(idLower).map(SeaShellConverter::toLowerDto).subscribeOn(parallel()),
                Pair::of).subscribeOn(parallel());
    }

    private Mono<SeaShellDto> fetchSeaShellPublisher(Shell seaShell) {
        final SeaShellDto seaShellDtoReturn = SeaShellConverter.toShellDto(seaShell);

        return zip(
                fetchPersonsPublisher(seaShell, seaShellDtoReturn).subscribeOn(Schedulers.parallel()),
                fetchCostumesPublisher(seaShell, seaShellDtoReturn).subscribeOn(Schedulers.parallel()),
                (persons, costumes) -> seaShellDtoReturn);
    }

    private Mono<?> fetchPersonsPublisher(Shell seaShell, SeaShellDto seaShellDtoReturn) {
        return from(from(just(
                seaShell.getPersons()).subscribeOn(parallel()).map(shellPersonRepository::findPersonsBlock).subscribeOn(parallel())
                .map(persons -> {
                    persons.forEach(person -> seaShellDtoReturn.getPersons().add(SeaShellConverter.toShellPersonDto(person)));
                    return seaShellDtoReturn.getPersons();
                }))
                .thenMany(zip(
                        fetchPersonAccountPublisher(seaShellDtoReturn).subscribeOn(parallel()),
                        fetchPersonFullCostumePublisher(seaShellDtoReturn).subscribeOn(parallel()))));

    }

    private Mono<SeaShellDto> fetchPersonFullCostumePublisher(SeaShellDto seaShellDtoReturn) {
        return from(fetchPersonCostumePublisher(seaShellDtoReturn)
                .thenMany(fetchPersonTopLowerPublisher(seaShellDtoReturn)));
    }

    private Mono<SeaShellDto> fetchPersonTopLowerPublisher(SeaShellDto seaShellDtoReturn) {
        return zip(
                fetchPersonCostumeTopPublisher(seaShellDtoReturn).subscribeOn(parallel()),
                fetchPersonCostumeLowerPublisher(seaShellDtoReturn).subscribeOn(parallel()),
                (persons, costumes) -> seaShellDtoReturn);
    }

    private Mono<?> fetchCostumesPublisher(Shell seaShell, SeaShellDto seaShellDtoReturn) {
        return from(just(seaShell.getCostumes()).subscribeOn(parallel()).map(shellCostumeRepository::findCostumesBlock).subscribeOn(parallel())
                .map(costumes -> {
                    costumes.forEach(costume -> seaShellDtoReturn.getCostumes().add(SeaShellConverter.toShellCostumeDto(costume)));
                    return costumes;
                })
                .thenMany(zip(
                        fetchCostumeTopPublisher(seaShellDtoReturn).subscribeOn(parallel()),
                        fetchCostumeLowerPublisher(seaShellDtoReturn).subscribeOn(parallel()),
                        (persons, costumes) -> seaShellDtoReturn)
                ));

    }

    private Mono<SeaShellDto> fetchCostumeLowerPublisher(SeaShellDto seaShellDtoReturn) {
        return fromCallable(() ->
                seaShellDtoReturn.getCostumes().parallelStream().map(seaShellCostumeDto ->
                {
                    seaShellCostumeDto.setLowerDto(
                            SeaShellConverter.toLowerDto(
                                    shellLowerRepository.findLowerByIdBlock(seaShellCostumeDto.getLowerId())));
                    return seaShellDtoReturn;
                }).findFirst().orElse(seaShellDtoReturn));
    }

    private Mono<SeaShellDto> fetchCostumeTopPublisher(SeaShellDto seaShellDtoReturn) {
        return fromCallable(() ->
                seaShellDtoReturn.getCostumes().parallelStream().map(seaShellCostumeDto ->
                {
                    seaShellCostumeDto.setTopDto(
                            SeaShellConverter.toTopDto(
                                    shellTopRepository.findTopByIdBlock(seaShellCostumeDto.getTopId())));
                    return seaShellDtoReturn;
                }).findFirst().orElse(seaShellDtoReturn));
    }

    private Mono<SeaShellDto> fetchPersonAccountPublisher(SeaShellDto seaShellDtoReturn) {
        return fromCallable(() ->
        {
            seaShellDtoReturn.getPersons()
                    .forEach(seaShellPersonDto -> seaShellPersonDto
                            .setAccountDto(
                                    SeaShellConverter
                                            .toAccountDto(shellAccountRepository
                                                    .findAccountByIdBlock(seaShellPersonDto.getAccountId()))));
            return seaShellDtoReturn;
        });
    }

    private Mono<SeaShellDto> fetchPersonCostumePublisher(SeaShellDto seaShellDtoReturn) {
        return fromCallable(() ->
        {
            seaShellDtoReturn.getPersons()
                    .forEach(seaShellPersonDto -> seaShellPersonDto
                            .setCostumeDto(
                                    SeaShellConverter
                                            .toShellCostumeDto(shellCostumeRepository
                                                    .findCostumeByIdBlock(seaShellPersonDto.getCostumeId()))));
            return seaShellDtoReturn;
        });
    }

    private Mono<SeaShellDto> fetchPersonCostumeLowerPublisher(SeaShellDto seaShellDtoReturn) {
        return fromCallable(() ->
                seaShellDtoReturn.getPersons().parallelStream().map(seaShellPersonDto -> {
                    SeaShellCostumeDto costumeDto = seaShellPersonDto.getCostumeDto();
                    costumeDto.setLowerDto(
                            SeaShellConverter.toLowerDto(
                                    shellLowerRepository.findLowerByIdBlock(costumeDto.getTopId())));
                    return seaShellDtoReturn;
                }).findFirst().orElse(seaShellDtoReturn));
    }

    private Mono<SeaShellDto> fetchPersonCostumeTopPublisher(SeaShellDto seaShellDtoReturn) {
        return fromCallable(() ->
                seaShellDtoReturn.getPersons().parallelStream().map(seaShellPersonDto -> {
                    SeaShellCostumeDto costumeDto = seaShellPersonDto.getCostumeDto();
                    costumeDto.setTopDto(
                            SeaShellConverter.toTopDto(
                                    shellTopRepository.findTopByIdBlock(costumeDto.getTopId())));
                    return seaShellDtoReturn;
                }).findFirst().orElse(seaShellDtoReturn));
    }
}

File: sea-shell-service-spring-web-flux/src/main/kotlin/org/jesperancinha/shell/webflux/service/SeaShellReactiveServiceImpl.kt
package org.jesperancinha.shell.webflux.service

import org.jesperancinha.shell.client.shells.Shell
import org.jesperancinha.shell.webflux.data.*
import org.jesperancinha.shell.webflux.repo.*
import org.springframework.data.util.Pair
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.publisher.ParallelFlux
import reactor.core.scheduler.Schedulers
import java.util.function.Consumer

@Service
class SeaShellReactiveServiceImpl(
    shellRepository: ShellRepositoryImpl,
    shellPersonRepository: ShellPersonRepositoryImpl,
    shellCostumeRepository: ShellCostumeRepositoryImpl,
    shellAccountRepository: ShellAccountRepositoryImpl,
    shellTopRepository: ShellTopRepositoryImpl,
    shellLowerRepository: ShellLowerRepositoryImpl
) : SeaShellOneAdapter(
    shellRepository,
    shellPersonRepository,
    shellCostumeRepository,
    shellAccountRepository,
    shellTopRepository,
    shellLowerRepository
) {
    fun allSeaShells(): ParallelFlux<SeaShellDto> = shellRepository
        .findAllSeaShells()
        .map { seaShell -> fetchSeaShellPublisher(seaShell) }
        .flatMap { source -> Flux.from(source) }

    fun getShell(id: Long): Mono<SeaShellDto> {
        return shellRepository.findSeaShellById(id)
            .map { seaShell -> fetchSeaShellPublisher(seaShell) }
            .flatMap { source -> Mono.from(source) }
    }

    fun getRootShell(idPerson: Long?, idCostume: Long): Mono<Pair<SeaShellPersonDto, SeaShellCostumeDto>> {
        return Mono.zip<SeaShellPersonDto?, SeaShellCostumeDto?, Pair<SeaShellPersonDto, SeaShellCostumeDto>>(
            shellPersonRepository.findPersonById(idPerson).map { SeaShellConverter.toShellPersonDto(it) }
                .subscribeOn(Schedulers.parallel()),
            shellCostumeRepository.findCostumeById(idCostume)
                .map { SeaShellConverter.toShellCostumeDto(it) }
                .subscribeOn(Schedulers.parallel())) { personDto, costumeDto ->
            Pair.of(
                personDto,
                costumeDto
            )
        }
            .subscribeOn(Schedulers.parallel())
    }

    fun getRootCostume(idTop: Long, idLower: Long): Mono<Pair<SeaShellTopDto, SeaShellLowerDto>> {
        return Mono.zip<SeaShellTopDto?, SeaShellLowerDto?, Pair<SeaShellTopDto, SeaShellLowerDto>>(
            shellTopRepository.findTopById(idTop).map { SeaShellConverter.toTopDto(it) }
                .subscribeOn(Schedulers.parallel()),
            shellLowerRepository.findLowerById(idLower).map { SeaShellConverter.toLowerDto(it) }
                .subscribeOn(Schedulers.parallel())) { topDto, lowerDto ->
            Pair.of(
                topDto,
                lowerDto
            )
        }
            .subscribeOn(Schedulers.parallel())
    }

    private fun fetchSeaShellPublisher(seaShell: Shell): Mono<SeaShellDto?> {
        val seaShellDtoReturn = SeaShellConverter.toShellDto(seaShell)
        return Mono.zip(
            fetchPersonsPublisher(seaShell, seaShellDtoReturn).subscribeOn(Schedulers.parallel()),
            fetchCostumesPublisher(seaShell, seaShellDtoReturn).subscribeOn(Schedulers.parallel())
        ) { _, _ -> seaShellDtoReturn }
    }

    private fun fetchPersonsPublisher(seaShell: Shell, seaShellDtoReturn: SeaShellDto): Mono<*> {
        return Mono.from(Mono.from(Mono.just(
            seaShell.persons.personId
        ).subscribeOn(Schedulers.parallel())
            .map { personIds: List<Long> -> shellPersonRepository.findPersonsBlock(personIds) }
            .subscribeOn(Schedulers.parallel())
            .map { persons ->
                persons.forEach { person ->
                    seaShellDtoReturn.persons.add(
                        SeaShellConverter.toShellPersonDto(person)
                    )
                }
                seaShellDtoReturn.persons
            })
            .thenMany(
                Mono.zip(
                    fetchPersonAccountPublisher(seaShellDtoReturn).subscribeOn(Schedulers.parallel()),
                    fetchPersonFullCostumePublisher(seaShellDtoReturn).subscribeOn(Schedulers.parallel())
                )
            )
        )
    }

    private fun fetchPersonFullCostumePublisher(seaShellDtoReturn: SeaShellDto): Mono<SeaShellDto?> {
        return Mono.from(
            fetchPersonCostumePublisher(seaShellDtoReturn)
                .thenMany(fetchPersonTopLowerPublisher(seaShellDtoReturn))
        )
    }

    private fun fetchPersonTopLowerPublisher(seaShellDtoReturn: SeaShellDto): Mono<SeaShellDto?> {
        return Mono.zip(
            fetchPersonCostumeTopPublisher(seaShellDtoReturn).subscribeOn(Schedulers.parallel()),
            fetchPersonCostumeLowerPublisher(seaShellDtoReturn).subscribeOn(Schedulers.parallel())
        ) { _, _ -> seaShellDtoReturn }
    }

    private fun fetchCostumesPublisher(seaShell: Shell, seaShellDtoReturn: SeaShellDto) = Mono.from(Mono.just(
        seaShell.costumes.costumeId
    ).subscribeOn(Schedulers.parallel())
        .map { costumeIds: List<Long> -> shellCostumeRepository.findCostumesBlock(costumeIds) }
        .subscribeOn(Schedulers.parallel())
        .map { costumes ->
            costumes.forEach(Consumer { costume ->
                seaShellDtoReturn.costumes.add(
                    SeaShellConverter.toShellCostumeDto(costume)
                )
            })
            costumes
        }
        .thenMany(Mono.zip(
            fetchCostumeTopPublisher(seaShellDtoReturn).subscribeOn(Schedulers.parallel()),
            fetchCostumeLowerPublisher(seaShellDtoReturn).subscribeOn(Schedulers.parallel())
        ) { _, _ -> seaShellDtoReturn }
        ))

    private fun fetchCostumeLowerPublisher(seaShellDtoReturn: SeaShellDto) = Mono.fromCallable {
        seaShellDtoReturn.costumes
            .map { seaShellCostumeDto ->
                seaShellCostumeDto?.copy(lowerDto = seaShellCostumeDto.lowerId
                    ?.let { shellLowerRepository.findLowerByIdBlock(it) }
                    ?.let { SeaShellConverter.toLowerDto(it) })
            }.let { seaShellDtoReturn.copy(costumes = it.toMutableList()) }
    }


    private fun fetchCostumeTopPublisher(seaShellDtoReturn: SeaShellDto) = Mono.fromCallable {
        seaShellDtoReturn.costumes
            .map { seaShellCostumeDto ->
                seaShellCostumeDto?.copy(topDto = seaShellCostumeDto.topId
                    ?.let { shellTopRepository.findTopByIdBlock(it) }
                    ?.let { SeaShellConverter.toTopDto(it) })
            }.let { seaShellDtoReturn.copy(costumes = it.toMutableList()) }
    }

    private fun fetchPersonAccountPublisher(seaShellDtoReturn: SeaShellDto) = Mono.fromCallable {
        seaShellDtoReturn.persons
            .map { seaShellPersonDto ->
                seaShellPersonDto?.copy(
                    accountDto = seaShellPersonDto.accountId
                        ?.let { shellAccountRepository.findAccountByIdBlock(it) }
                        ?.let { SeaShellConverter.toAccountDto(it) }
                )
            }.let { seaShellDtoReturn.copy(persons = it.toMutableList()) }
    }

    private fun fetchPersonCostumePublisher(seaShellDtoReturn: SeaShellDto): Mono<SeaShellDto?> {
        return Mono.fromCallable {
            seaShellDtoReturn.persons
                .map { seaShellPersonDto: SeaShellPersonDto? ->
                    seaShellPersonDto?.copy(costumeDto = seaShellPersonDto.costumeId
                        ?.let { shellCostumeRepository.findCostumeByIdBlock(it) }
                        ?.let { SeaShellConverter.toShellCostumeDto(it) })
                }.let { seaShellDtoReturn.copy(persons = it.toMutableList()) }

        }
    }

    private fun fetchPersonCostumeLowerPublisher(seaShellDtoReturn: SeaShellDto): Mono<SeaShellDto?> {
        return Mono.fromCallable {
            seaShellDtoReturn.persons
                .map { seaShellPersonDto ->
                    val costumeDto = seaShellPersonDto?.costumeDto
                    seaShellPersonDto?.copy(
                        costumeDto = costumeDto?.copy(
                            lowerDto = costumeDto.topId
                                ?.let { shellLowerRepository.findLowerByIdBlock(it) }
                                ?.let {
                                    SeaShellConverter.toLowerDto(it)
                                }
                        )
                    )
                }.let { seaShellDtoReturn.copy(persons = it.toMutableList()) }
        }
    }

    private fun fetchPersonCostumeTopPublisher(seaShellDtoReturn: SeaShellDto): Mono<SeaShellDto?> {
        return Mono.fromCallable {
            seaShellDtoReturn.persons
                .map { seaShellPersonDto ->
                    val costumeDto = seaShellPersonDto?.costumeDto
                    seaShellPersonDto?.copy(costumeDto = costumeDto?.copy(topDto = costumeDto.topId
                        ?.let { shellTopRepository.findTopByIdBlock(it) }
                        ?.let {
                            SeaShellConverter.toTopDto(
                                it
                            )
                        })
                    )
                }.let { seaShellDtoReturn.copy(persons = it.toMutableList()) }
        }
    }
}